File: sdk/c/src/providers/temporary_email_org.c
/**
 * temporary-email.org 渠道
 * 首次 GET /zh/messages 下发 Cookie；收信需 Cookie + x-requested-with
 */
#include "tempmail_internal.h"
#include <string.h>

#define TEO_MESSAGES "https://www.temporary-email.org/zh/messages"
#define TEO_REFERER "https://www.temporary-email.org/zh"

static const char* teo_headers_base[] = {
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
    "Accept: text/plain, */*; q=0.01",
    "accept-language: zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "cache-control: no-cache",
    "dnt: 1",
    "pragma: no-cache",
    "priority: u=1, i",
    "referer: " TEO_REFERER,
    "sec-ch-ua: \"Chromium\";v=\"146\", \"Not-A.Brand\";v=\"24\", \"Microsoft Edge\";v=\"146\"",
    "sec-ch-ua-mobile: ?0",
    "sec-ch-ua-platform: \"Windows\"",
    "sec-fetch-dest: empty",
    "sec-fetch-mode: cors",
    "sec-fetch-site: same-origin",
    NULL
};

tm_email_info_t* tm_provider_temporary_email_org_generate(void) {
    tm_http_response_t *resp = tm_http_request(TM_HTTP_GET, TEO_MESSAGES, teo_headers_base, NULL, 15);
    if (!resp || resp->status != 200) { tm_http_response_free(resp); return NULL; }

    if (!resp->cookies || !strstr(resp->cookies, "temporaryemail_session=") || !strstr(resp->cookies, "email=")) {
        tm_http_response_free(resp);
        return NULL;
    }

    char *cookie_copy = tm_strdup(resp->cookies);
    cJSON *json = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!json) { free(cookie_copy); return NULL; }

    const char *mailbox = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(json, "mailbox"), "");
    if (!mailbox[0] || !strchr(mailbox, '@')) {
        cJSON_Delete(json);
        free(cookie_copy);
        return NULL;
    }

    tm_email_info_t *info = tm_email_info_new();
    info->channel = CHANNEL_TEMPORARY_EMAIL_ORG;
    info->email = tm_strdup(mailbox);
    info->token = cookie_copy;
    cJSON_Delete(json);
    return info;
}

tm_email_t* tm_provider_temporary_email_org_get_emails(const char *token, const char *email, int *count) {
    *count = -1;
    if (!token || !email) return NULL;

    char cookie_hdr[6144];
    if (snprintf(cookie_hdr, sizeof(cookie_hdr), "Cookie: %s", token) >= (int)sizeof(cookie_hdr)) {
        return NULL;
    }

    const char* headers[] = {
        cookie_hdr,
        "x-requested-with: XMLHttpRequest",
        teo_headers_base[0],
        teo_headers_base[1],
        teo_headers_base[2],
        teo_headers_base[3],
        teo_headers_base[4],
        teo_headers_base[5],
        teo_headers_base[6],
        teo_headers_base[7],
        teo_headers_base[8],
        teo_headers_base[9],
        teo_headers_base[10],
        teo_headers_base[11],
        teo_headers_base[12],
        teo_headers_base[13],
        NULL
    };

    tm_http_response_t *resp = tm_http_request(TM_HTTP_GET, TEO_MESSAGES, headers, NULL, 15);
    if (!resp || resp->status != 200) { tm_http_response_free(resp); return NULL; }

    cJSON *json = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!json) return NULL;

    cJSON *arr = cJSON_GetObjectItemCaseSensitive(json, "messages");
    int n = cJSON_IsArray(arr) ? cJSON_GetArraySize(arr) : 0;
    *count = n;
    if (n == 0) { cJSON_Delete(json); return NULL; }

    tm_email_t *emails = tm_emails_new(n);
    for (int i = 0; i < n; i++) emails[i] = tm_normalize_email(cJSON_GetArrayItem(arr, i), email);
    cJSON_Delete(json);
    return emails;
}

File: sdk/rust/src/providers/temporary_email_org.rs
/*!
 * temporary-email.org 渠道
 * 首次 GET /zh/messages 下发 Cookie；收信需带 Cookie 与 X-Requested-With: XMLHttpRequest
 */

use serde_json::Value;
use crate::types::{Channel, EmailInfo, Email};
use crate::normalize::normalize_email;
use crate::config::{http_client, block_on, get_current_ua};

const MESSAGES_URL: &str = "https://www.temporary-email.org/zh/messages";
const REFERER: &str = "https://www.temporary-email.org/zh";

pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let resp = http_client()
            .get(MESSAGES_URL)
            .header("User-Agent", get_current_ua())
            .header("Accept", "text/plain, */*; q=0.01")
            .header("accept-language", "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6")
            .header("cache-control", "no-cache")
            .header("dnt", "1")
            .header("pragma", "no-cache")
            .header("priority", "u=1, i")
            .header("referer", REFERER)
            .header("sec-ch-ua", r#""Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146""#)
            .header("sec-ch-ua-mobile", "?0")
            .header("sec-ch-ua-platform", r#""Windows""#)
            .header("sec-fetch-dest", "empty")
            .header("sec-fetch-mode", "cors")
            .header("sec-fetch-site", "same-origin")
            .send().await.map_err(|e| format!("temporary-email-org request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("temporary-email-org generate failed: {}", resp.status()));
        }

        let cookie: String = resp.headers().get_all("set-cookie")
            .iter()
            .filter_map(|v| v.to_str().ok())
            .filter_map(|s| s.split(';').next().map(|p| p.trim().to_string()))
            .filter(|p| !p.is_empty())
            .collect::<Vec<_>>()
            .join("; ");
        if cookie.is_empty() {
            return Err("temporary-email-org: no set-cookie".into());
        }
        if !cookie.contains("temporaryemail_session=") || !cookie.contains("email=") {
            return Err("temporary-email-org: missing session cookies".into());
        }

        let data: Value = resp.json().await.map_err(|e| format!("parse failed: {}", e))?;
        let mailbox = data["mailbox"].as_str().unwrap_or("").trim();
        if mailbox.is_empty() || !mailbox.contains('@') {
            return Err("temporary-email-org: invalid mailbox".into());
        }

        Ok(EmailInfo {
            channel: Channel::TemporaryEmailOrg,
            email: mailbox.to_string(),
            token: Some(cookie),
            expires_at: None,
            created_at: None,
        })
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let token = token.to_string();
    let email = email.to_string();
    block_on(async {
        let resp = http_client()
            .get(MESSAGES_URL)
            .header("User-Agent", get_current_ua())
            .header("Accept", "text/plain, */*; q=0.01")
            .header("accept-language", "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6")
            .header("cache-control", "no-cache")
            .header("dnt", "1")
            .header("pragma", "no-cache")
            .header("priority", "u=1, i")
            .header("referer", REFERER)
            .header("sec-ch-ua", r#""Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146""#)
            .header("sec-ch-ua-mobile", "?0")
            .header("sec-ch-ua-platform", r#""Windows""#)
            .header("sec-fetch-dest", "empty")
            .header("sec-fetch-mode", "cors")
            .header("sec-fetch-site", "same-origin")
            .header("x-requested-with", "XMLHttpRequest")
            .header("Cookie", &token)
            .send().await.map_err(|e| format!("temporary-email-org request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("temporary-email-org get emails failed: {}", resp.status()));
        }

        let data: Value = resp.json().await.map_err(|e| format!("parse failed: {}", e))?;
        let arr = data["messages"].as_array().cloned().unwrap_or_default();
        Ok(arr.iter().map(|raw| normalize_email(raw, &email)).collect())
    })
}
